device/cuda/src/ambiguity_resolution/kernels/rearrange_tracks.cu

```diff
@@ -68,116 +79,140 @@ __global__ void rearrange_tracks(device::rearrange_tracks_payload payload) {
+                                (rel_sh == rel_sh_ref && pval >= pval_ref)) {
+
+                                left = mid + 1;
+                            } else {
+                                right = mid;
+                            }
+                        }
                     }
-                }
-            }
 
-            int delta = 0;
+                    int delta = delta =
+                        gid - left - (prefix_sums[gid] - prefix_sums[left]);
 
-            if (is_updated[sorted_ids[left]]) {
-                delta = gid - left - (prefix_sums[gid] - prefix_sums[left]);
-            } else {
-                delta = gid - left - (prefix_sums[gid] - prefix_sums[left] - 1);
-            }
+                    if (!is_updated[sorted_ids[left]]) {
+                        delta++;
+                    }
 
-            shifted_idx -= delta;
-        }
+                    atomicAdd(&shifted_idx, -delta);
+                }
+            }
 
-        for (int i = 0; i < N; i++) {
+            for (int i = ini_idx; i < fin_idx; i++) {
 
-            auto id = updated_tracks[i];
+                auto id = updated_tracks[i];
 
-            if (inverted_ids[id] < gid) {
-                shifted_idx--;
+                if (inverted_ids[id] < gid) {
+                    atomicAdd(&shifted_idx, -1);
+                }
             }
-        }
 
-        int offset = 0;
-        for (int i = 0; i < N; i++) {
-            if (updated_tracks[i] == tid) {
-                offset = i;
-                break;
+            int offset = 0;
+            for (int i = ini_idx; i < fin_idx; i++) {
+                if (updated_tracks[i] == tid) {
+                    offset = i;
+                    break;
+                }
             }
-        }
-        shifted_idx += offset;
+            if (offset != 0) {
+                atomicAdd(&shifted_idx, offset);
+            }
+        } else {
 
-    } else {
-        for (int i = 0; i < N; i++) {
+            for (int i = ini_idx; i < fin_idx; i++) {
 
-            auto id = updated_tracks[i];
-            auto rel_sh = rel_shared[id];
-            auto pval = pvals[id];
+                auto id = updated_tracks[i];
+                auto rel_sh = rel_shared[id];
+                auto pval = pvals[id];
 
-            if (inverted_ids[id] > gid) {
-                if (rel_sh < rel_sh_ref) {
-                    shifted_idx++;
-                } else if (rel_sh == rel_sh_ref && pval > pval_ref) {
-                    shifted_idx++;
+                if (inverted_ids[id] > gid) {
+                    if (rel_sh < rel_sh_ref) {
+                        atomicAdd(&shifted_idx, 1);
+                    } else if (rel_sh == rel_sh_ref && pval > pval_ref) {
+                        atomicAdd(&shifted_idx, 1);
+                    }
                 }
             }
         }
     }
 
-    temp_sorted_ids.at(shifted_idx) = tid;
+    __syncthreads();
```

Comment:
Technically `__syncthreads()` is not supposed to be here as the threads in the same warp work on the same thread. However, I found that the tests fail with CUDA 12.6 (CI) while they pass with CUDA 12.4. It seems CUDA 12.6 requires synchronization with atomicAdd even for the threads in the same warp. Not sure if this is a buggy behavior or an intended change, but I don't care as long as it works